csfieldguide/chapters/content/en/coding-compression/sections/general-purpose.md

```diff
@@ -43,27 +31,340 @@ Here are some of the bits from the example earlier in this chapter; you can past
+|------|------------------|
+| 0    | I                |
+| 1    | AM               |
+| 2    | SAM              |
+| 3    | ,                |
+
+
+If you receive the code `0 1 2 3 2 0 1`, you could look up each number in the "dictionary", and translate it to `I AM SAM, SAM I AM`.
+Of course, data on a computer is represented as binary numbers; with 4 entries in the dictionary, you'd need 2 bits for each number, and the code would actually be `00 01 10 11 10 00 01` and the dictionary would look like this:
+
+| Code (in binary) | Character String |
+|------------------|------------------|
+| 00               | I                |
+| 01               | AM               |
+| 10               | SAM              |
+| 11               | ,                |
+
+This dictionary is way too small for regular text.
+Suppose you were allowed to use a dictionary of 4096 strings of characters instead of just 4.
+Which strings of characters would you put in the dictionary to make sure it could represent any possible text?
+And how many bits are needed to give the index number in the dictionary?
+
+You can try creating your own dictionary for some text in the following interactive &ndash; see how few dictionary entries you need to cover as much of the text as possible.
+
+{interactive slug="dictionary-compression" type="whole-page"}
 
 {panel type="teacher-note"}
 
-# Unplugged activity on Ziv-Lempel approach
+# Teacher note
+
+Students might think of putting in common words like “the” and “of”.
+Or even better, the strings might include a space after them  “the “ and “of “. But to be sure of representing any text, they’ll also need to include letters of the alphabet, and other symbols.
+If given long strings of text, such as "as soon as", then large amounts of text can be replaced with one code, but the longer strings might not occur a lot.
+Choosing a dictionary is hard work!
+
+{panel end}
+
+
+Choosing good entries for a dictionary is hard; the LZW method that we're about to look at has a clever way of making a dictionary that suits the particular language being compressed, and it builds the dictionary while the file is being compressed!
+
+The interactive below demonstrates LZW compression.
+You might want to have it open in a separate window to try things out while you read how it works below.
+
+{interactive slug="lzw-compression" parameters="message-length=short" type="whole-page"}
+
+## LZW Encoding
+
+Initially the LZW encoder begins with a dictionary that can store up to 4096 strings of characters.
+Codes 0 to 255 represent single character sequences (the alphabet, punctuation, etc).
+The codes will need to be stored as binary numbers, therefore each code needs 12 bits to represent it in order to represent a total of 4096 entries (since 12 bits allows us to represent 4096 different values).
+
+As the message is being compressed, at each step the encoder adds a new string of characters to the dictionary (filling in dictionary spots 255 &ndash; 4095).
```

Comment:
This one doesn't need an ndash &ndash; the ndash is used like for what I'm doing now, for supplemental information